Introduction
============

The aim of this library [#]_ is to provide a series of useful modules and packages to make it easier to build custom pipelines for metagenomics or any kind of bioinformatics analysis. It integrates other well known python libraries in bioinformatics, like `HTSeq <http://www-huber.embl.de/users/anders/HTSeq/>`_, `pysam <https://code.google.com/p/pysam/>`_, `numpy <http://www.numpy.org>`_ and `scipy <http://www.scipy.org>`_.

A tutorial pipeline is provided in the `Documentation <http://pythonhosted.org//mgkit/pipeline/tutorial.html>`_.

A discussion mailing list is available at `mgkit-users <https://groups.google.com/forum/#!forum/mgkit-users>`_.

Citing
------

Rubino, F. and Creevey, C.J. 2014. MGkit: Metagenomic Framework For The Study Of Microbial Communities. . Available at: `figshare <http://figshare.com/articles/MGkit_Metagenomic_Framework_For_The_Study_Of_Microbial_Communities/1269288>`_ [doi:10.6084/m9.figshare.1269288].`

a citation is also available using the :func:`mgkit.cite()` function or using the `--cite` option on all scripts included. For example:

.. code-block:: bash

    blast2gff --cite

Links
-----

.. [#] `<https://bitbucket.org/setsuna80/mgkit>`_
